cpp/src/arrow/compute/kernels/aggregate_basic_internal.h

```diff
@@ -49,20 +50,22 @@ void AddMinMaxKernel(KernelInit init, internal::detail::GetTypeId get_id,
 
 // SIMD variants for kernels
 void AddSumAvx2AggKernels(ScalarAggregateFunction* func);
+void AddSumCheckedAvx2AggKernels(ScalarAggregateFunction* func);
 void AddMeanAvx2AggKernels(ScalarAggregateFunction* func);
 void AddMinMaxAvx2AggKernels(ScalarAggregateFunction* func);
 
 void AddSumAvx512AggKernels(ScalarAggregateFunction* func);
+void AddSumCheckedAvx512AggKernels(ScalarAggregateFunction* func);
 void AddMeanAvx512AggKernels(ScalarAggregateFunction* func);
 void AddMinMaxAvx512AggKernels(ScalarAggregateFunction* func);
 
 // ----------------------------------------------------------------------
 // Sum implementation
 
-template <typename ArrowType, SimdLevel::type SimdLevel,
+template <typename ArrowType, SimdLevel::type SimdLevel, bool Checked,
```

Comment:
TBH, I'm not sure `Checked` and `SimdLevel` need to be template parameters at this level. `SumArray` is performance critical, but the orchestration in `SumImpl` probably much less so.

Not making those template parameters may save a bit on code generation and compilation times. Perhaps open an issue about that?

Reply:
I created #37585 for this